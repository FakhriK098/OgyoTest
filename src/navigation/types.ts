import { NavigatorScreenParams } from '@react-navigation/native';

export type RootStackParamList = {
  MainTabs: NavigatorScreenParams<MainTabParamList>;
  Detail: { id: string; title: string; data?: any };
};

export type HomeStackParamList = {
  Detail: { id: string };
};

export type MainTabParamList = {
  Home: undefined;
  User: undefined;
};
