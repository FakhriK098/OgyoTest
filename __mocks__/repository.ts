import { IRepository } from 'src/types/repository';

export const dummyRepositories: IRepository[] = [
  {
    id: 1,
    node_id: 'MDEwOlJlcG9zaXRvcnkx',
    name: 'hello-world',
    full_name: 'octocat/hello-world',
    private: false,
    owner: {
      login: 'octocat',
      id: 1,
      node_id: 'MDQ6VXNlcjE=',
      avatar_url: 'https://avatars.githubusercontent.com/u/1?v=4',
      gravatar_id: '',
      url: 'https://api.github.com/users/octocat',
      html_url: 'https://github.com/octocat',
      followers_url: 'https://api.github.com/users/octocat/followers',
      following_url:
        'https://api.github.com/users/octocat/following{/other_user}',
      gists_url: 'https://api.github.com/users/octocat/gists{/gist_id}',
      starred_url:
        'https://api.github.com/users/octocat/starred{/owner}{/repo}',
      subscriptions_url: 'https://api.github.com/users/octocat/subscriptions',
      organizations_url: 'https://api.github.com/users/octocat/orgs',
      repos_url: 'https://api.github.com/users/octocat/repos',
      events_url: 'https://api.github.com/users/octocat/events{/privacy}',
      received_events_url:
        'https://api.github.com/users/octocat/received_events',
      type: 'User',
      user_view_type: 'public',
      site_admin: false,
    },
    html_url: 'https://github.com/octocat/hello-world',
    description: 'A simple hello world repository',
  },
  {
    id: 2,
    node_id: 'MDEwOlJlcG9zaXRvcnky',
    name: 'awesome-project',
    full_name: 'johndoe/awesome-project',
    private: false,
    owner: {
      login: 'johndoe',
      id: 2,
      node_id: 'MDQ6VXNlcjI=',
      avatar_url: 'https://avatars.githubusercontent.com/u/2?v=4',
      gravatar_id: '',
      url: 'https://api.github.com/users/johndoe',
      html_url: 'https://github.com/johndoe',
      followers_url: 'https://api.github.com/users/johndoe/followers',
      following_url:
        'https://api.github.com/users/johndoe/following{/other_user}',
      gists_url: 'https://api.github.com/users/johndoe/gists{/gist_id}',
      starred_url:
        'https://api.github.com/users/johndoe/starred{/owner}{/repo}',
      subscriptions_url: 'https://api.github.com/users/johndoe/subscriptions',
      organizations_url: 'https://api.github.com/users/johndoe/orgs',
      repos_url: 'https://api.github.com/users/johndoe/repos',
      events_url: 'https://api.github.com/users/johndoe/events{/privacy}',
      received_events_url:
        'https://api.github.com/users/johndoe/received_events',
      type: 'User',
      user_view_type: 'public',
      site_admin: false,
    },
    html_url: 'https://github.com/johndoe/awesome-project',
    description: 'An awesome project to test API responses',
  },
  {
    id: 3,
    node_id: 'MDEwOlJlcG9zaXRvcnkz',
    name: 'cool-library',
    full_name: 'janedoe/cool-library',
    private: true,
    owner: {
      login: 'janedoe',
      id: 3,
      node_id: 'MDQ6VXNlcjM=',
      avatar_url: 'https://avatars.githubusercontent.com/u/3?v=4',
      gravatar_id: '',
      url: 'https://api.github.com/users/janedoe',
      html_url: 'https://github.com/janedoe',
      followers_url: 'https://api.github.com/users/janedoe/followers',
      following_url:
        'https://api.github.com/users/janedoe/following{/other_user}',
      gists_url: 'https://api.github.com/users/janedoe/gists{/gist_id}',
      starred_url:
        'https://api.github.com/users/janedoe/starred{/owner}{/repo}',
      subscriptions_url: 'https://api.github.com/users/janedoe/subscriptions',
      organizations_url: 'https://api.github.com/users/janedoe/orgs',
      repos_url: 'https://api.github.com/users/janedoe/repos',
      events_url: 'https://api.github.com/users/janedoe/events{/privacy}',
      received_events_url:
        'https://api.github.com/users/janedoe/received_events',
      type: 'User',
      user_view_type: 'public',
      site_admin: false,
    },
    html_url: 'https://github.com/janedoe/cool-library',
    description: 'A private library for internal usage',
  },
];
