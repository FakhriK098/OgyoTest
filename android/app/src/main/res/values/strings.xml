<resources>
    <string name="app_name">OgyoTest</string>
</resources>
