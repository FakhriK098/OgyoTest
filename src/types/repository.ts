export interface IOwner {
  login: string;
  id: number;
  node_id: string;
  avatar_url: string;
  gravatar_id: string;
  url: string;
  html_url: string;
  followers_url: string;
  following_url: string;
  gists_url: string;
  starred_url: string;
  subscriptions_url: string;
  organizations_url: string;
  repos_url: string;
  events_url: string;
  received_events_url: string;
  type: string;
  user_view_type: string;
  site_admin: boolean;
}

export interface IRepository {
  id: number;
  node_id: string;
  name: string;
  full_name: string;
  private: boolean;
  owner: IOwner;
  html_url: string;
  description: string;
}

export interface IFetchRepositoriesPayload {
  since?: number;
}

export interface ISearchRepositoriesPayload {
  query: string;
}

export interface ILinkHeader {
  next?: string;
  prev?: string;
  first?: string;
  last?: string;
}
