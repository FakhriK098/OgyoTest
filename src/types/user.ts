export interface IUser {
  id: number;
  node_id: string;
  name: string;
  full_name: string;
  private: boolean;
  owner: IUserOwner;
  html_url: string;
  description: string;
  size: number;
  stargazers_count: number;
  watchers_count: number;
  language: string;
  forks_count: number;
  open_issues_count: number;
  license: ILicense;
  topics: string[];
  network_count: number;
  subscribers_count: number;
}

export interface ILicense {
  key: string;
  name: string;
  spdx_id: string;
  url: string;
  node_id: string;
}

export interface IUserOwner {
  login: string;
  id: number;
  node_id: string;
  avatar_url: string;
  gravatar_id: string;
  url: string;
  html_url: string;
  followers_url: string;
  following_url: string;
  gists_url: string;
  starred_url: string;
  subscriptions_url: string;
  organizations_url: string;
  repos_url: string;
  events_url: string;
  received_events_url: string;
  type: string;
  user_view_type: string;
  site_admin: boolean;
}

export interface IFetchUserPayload {
  fullName?: string;
}

export interface ILabelProps {
  title: string;
  value?: string;
}
