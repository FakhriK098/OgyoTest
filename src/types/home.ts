import { IRepository } from './repository';

export interface IFloatingActionButtonProps {
  onPress: () => void;
}

export interface ISearchBarProps {
  onSearch: (query: string) => void;
  onSelectSuggestion: (repository: IRepository) => void;
  onClear: () => void;
  suggestions: IRepository[];
  loading: boolean;
  placeholder?: string;
  onSubmitSearch: (query: string) => void;
}

export type TSortOrder = 'ASC' | 'DESC';

export interface ISortFilterModalProps {
  visible: boolean;
  onClose: () => void;
  onSort: (order: TSortOrder) => void;
  currentOrder: TSortOrder;
}
