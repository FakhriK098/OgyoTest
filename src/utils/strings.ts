import { ILinkHeader } from '../types/repository';

export const getFullName = (name: string): string => {
  return name.split('/').join(' ');
};

export const parseLinkHeader = (linkHeader?: string): ILinkHeader => {
  const links: ILinkHeader = {};
  if (!linkHeader) return links;

  const parts = linkHeader.split(',');
  parts.forEach(part => {
    const match = part.match(/<([^>]+)>;\s*rel="([^"]+)"/);
    if (match) {
      const [, url, rel] = match;
      links[rel as keyof ILinkHeader] = url;
    }
  });

  return links;
};

export const extractSinceFromUrl = (url?: string): number | null => {
  if (!url) return null;
  const match = url.match(/since=(\d+)/);
  return match ? parseInt(match[1], 10) : null;
};
