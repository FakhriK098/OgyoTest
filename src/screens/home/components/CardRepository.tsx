import { StyleSheet, Text, View } from 'react-native';
import React from 'react';
import { IRepository } from '../../../types/repository';
import { colors } from '../../../themes/colors';
import { getFullName } from '../../../utils/strings';

const CardRepository = ({ item }: { item: IRepository }) => {
  return (
    <View style={styles.container}>
      <Text style={styles.title}>{item.name}</Text>
      <Text style={styles.name}>{getFullName(item.full_name)}</Text>
    </View>
  );
};

export default CardRepository;

const styles = StyleSheet.create({
  container: {
    backgroundColor: colors.white,
    borderRadius: 8,
    padding: 8,
  },
  title: {
    fontSize: 16,
    fontWeight: 'medium',
    color: colors.black,
  },
  name: {
    fontSize: 8,
    fontWeight: 'light',
    color: colors.shade800,
    marginTop: 4,
  },
});
