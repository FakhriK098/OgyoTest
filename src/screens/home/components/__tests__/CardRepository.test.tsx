import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import CardRepository from '../CardRepository';
import { IRepository } from 'src/types/repository';
import * as navigation from '@react-navigation/native';
import * as stringUtils from 'src/utils/strings';

// Mock dependencies
jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn(),
}));

jest.mock('src/utils/strings', () => ({
  getFullName: jest.fn(),
}));

const mockNavigate = jest.fn();
const mockNavigation = {
  navigate: mockNavigate,
  goBack: jest.fn(),
  canGoBack: jest.fn(() => true),
  dispatch: jest.fn(),
  getId: jest.fn(),
  getParent: jest.fn(),
  getState: jest.fn(),
  isFocused: jest.fn(() => true),
  removeListener: jest.fn(),
  reset: jest.fn(),
  setOptions: jest.fn(),
  setParams: jest.fn(),
  addListener: jest.fn(),
};

const mockRepository: IRepository = {
  id: 1,
  node_id: 'MDEwOlJlcG9zaXRvcnkx',
  name: 'react',
  full_name: 'facebook/react',
  private: false,
  description:
    'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
  html_url: 'https://github.com/facebook/react',
  owner: {
    login: 'facebook',
    id: 69631,
    node_id: 'MDEyOk9yZ2FuaXphdGlvbjY5NjMx',
    avatar_url: 'https://avatars.githubusercontent.com/u/69631?v=4',
    gravatar_id: '',
    url: 'https://api.github.com/users/facebook',
    html_url: 'https://github.com/facebook',
    followers_url: 'https://api.github.com/users/facebook/followers',
    following_url:
      'https://api.github.com/users/facebook/following{/other_user}',
    gists_url: 'https://api.github.com/users/facebook/gists{/gist_id}',
    starred_url: 'https://api.github.com/users/facebook/starred{/owner}{/repo}',
    subscriptions_url: 'https://api.github.com/users/facebook/subscriptions',
    organizations_url: 'https://api.github.com/users/facebook/orgs',
    repos_url: 'https://api.github.com/users/facebook/repos',
    events_url: 'https://api.github.com/users/facebook/events{/privacy}',
    received_events_url:
      'https://api.github.com/users/facebook/received_events',
    type: 'Organization',
    user_view_type: 'public',
    site_admin: false,
  },
};

describe('CardRepository', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    (navigation.useNavigation as jest.Mock).mockReturnValue(mockNavigation);
    (stringUtils.getFullName as jest.Mock).mockImplementation(
      (fullName: string) => fullName.split('/').join(' '),
    );
  });

  it('should render repository information correctly', () => {
    const { getByText } = render(<CardRepository item={mockRepository} />);

    expect(getByText('react')).toBeTruthy();
    expect(getByText('facebook react')).toBeTruthy();
  });

  it('should handle press event and navigate to detail screen', () => {
    const { getByText } = render(<CardRepository item={mockRepository} />);

    // The component is wrapped in a Pressable, so we can press the text element's parent
    const titleElement = getByText('react');
    const card = titleElement.parent;

    if (card) {
      fireEvent.press(card);

      expect(mockNavigate).toHaveBeenCalledWith('Detail', {
        id: 'facebook/react',
      });
    }
  });

  it('should render with different repository data', () => {
    const differentRepo: IRepository = {
      ...mockRepository,
      id: 2,
      name: 'vue',
      full_name: 'vuejs/vue',
      description: 'Vue.js is a progressive framework',
      owner: {
        ...mockRepository.owner,
        login: 'vuejs',
        id: 123456,
      },
    };

    (stringUtils.getFullName as jest.Mock).mockImplementation(
      (fullName: string) => fullName.split('/').join(' '),
    );

    const { getByText } = render(<CardRepository item={differentRepo} />);

    expect(getByText('vue')).toBeTruthy();
    expect(getByText('vuejs vue')).toBeTruthy();
    expect(stringUtils.getFullName).toHaveBeenCalledWith('vuejs/vue');
  });

  it('should handle press event with different repository id', () => {
    const differentRepo: IRepository = {
      ...mockRepository,
      id: 2,
      name: 'angular',
      full_name: 'angular/angular',
    };

    const { getByText } = render(<CardRepository item={differentRepo} />);

    const card = getByText('angular').parent;
    if (card) {
      fireEvent.press(card);

      expect(mockNavigate).toHaveBeenCalledWith('Detail', {
        id: 'angular/angular',
      });
    }
  });

  it('should be pressable and have correct accessibility properties', () => {
    const { getByText, UNSAFE_getByType } = render(
      <CardRepository item={mockRepository} />,
    );

    const Pressable = require('react-native').Pressable;
    const pressable = UNSAFE_getByType(Pressable);
    expect(pressable).toBeTruthy();

    expect(getByText('react')).toBeTruthy();
  });

  // it('should render with correct styles', () => {
  //   const { getByText } = render(<CardRepository item={mockRepository} />);

  //   const titleText = getByText('react');
  //   const nameText = getByText('facebook react');

  //   // Check that text elements exist and can be styled
  //   expect(titleText).toBeTruthy();
  //   expect(nameText).toBeTruthy();

  //   // Verify the text content is correct
  //   expect(titleText.props.children).toBe('react');
  //   expect(nameText.props.children).toBe('facebook react');
  // });

  it('should handle navigation errors gracefully', () => {
    const consoleSpy = jest.spyOn(console, 'error').mockImplementation();
    mockNavigate.mockImplementationOnce(() => {
      throw new Error('Navigation failed');
    });

    const { getByText } = render(<CardRepository item={mockRepository} />);

    const card = getByText('react').parent;
    if (card) {
      expect(() => fireEvent.press(card)).toThrow('Navigation failed');
    }

    consoleSpy.mockRestore();
  });

  it('should match snapshot', () => {
    const { toJSON } = render(<CardRepository item={mockRepository} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should match snapshot with different repository data', () => {
    const differentRepo: IRepository = {
      ...mockRepository,
      id: 5,
      name: 'test-repo',
      full_name: 'test-org/test-repo',
    };

    (stringUtils.getFullName as jest.Mock).mockImplementation(
      (fullName: string) => fullName.split('/').join(' '),
    );

    const { toJSON } = render(<CardRepository item={differentRepo} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
