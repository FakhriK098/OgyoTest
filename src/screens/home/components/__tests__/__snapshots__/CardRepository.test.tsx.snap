// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardRepository should match snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 8,
      "padding": 8,
    }
  }
>
  <Text
    style={
      {
        "color": "#000000",
        "fontSize": 16,
        "fontWeight": "medium",
      }
    }
  >
    react
  </Text>
  <Text
    style={
      {
        "color": "#161B28",
        "fontSize": 8,
        "fontWeight": "light",
        "marginTop": 4,
      }
    }
  >
    facebook react
  </Text>
</View>
`;

exports[`CardRepository should match snapshot with different repository data 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 8,
      "padding": 8,
    }
  }
>
  <Text
    style={
      {
        "color": "#000000",
        "fontSize": 16,
        "fontWeight": "medium",
      }
    }
  >
    test-repo
  </Text>
  <Text
    style={
      {
        "color": "#161B28",
        "fontSize": 8,
        "fontWeight": "light",
        "marginTop": 4,
      }
    }
  >
    test-org test-repo
  </Text>
</View>
`;
